src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,277 @@
+			r.logger.Error(err, "Failed to update RedpandaClusterStatus")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string,
+) error {
+	r.logger.Info("Start statefulset image update")
+
+	// When a StatefulSet's partition number is set to `i`, only Pods with ordinal
+	// greater than or equal to `i` will be updated.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#partitions
+	for ordinal := *sts.Spec.Replicas - 1; ordinal >= 0; ordinal-- {
+		// Update() on statefulset has not been called yet in this run, however,
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
+			continue
+		}
+
+		// Continue only if error is due to pod not ready, or unchanged image.
+		if !errors.Is(poderr, errContainerHasWrongImage) && !errors.Is(poderr, errPodNotReady) {
+			return poderr
+		}
+
+		// Ideally, at this point we'd wait until all Pods are in ready state.
+		// However, if something went wrong with a previous image change and the pod
+		// is in bad state, e.g., due to wrong or erroneous image, the pod will
+		// keep restarting and require manual intervention. Instead, for now we
+		// skip the readiness requirement and "force" the image update.
+
+		if err := r.rollingUpdatePartition(ctx, sts, ordinal); err != nil {
+			return err
+		}
+
+		poderr = r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr != nil {
+			msg := fmt.Sprintf("Pod not ready, will retry - ordinal %d", ordinal)
+			r.logger.Info(msg)
+
+			return &NeedToReconcileError{RequeueAfter: requeueDuration, Msg: msg}
+		}
+
+		// Give time to redpanda pod to rejoin cluster.
+		// TODO: In a revision, an API call to redpanda is made to help us decide when to continue.
+		// If the answer is negative, we would requeue as above.
+		r.logger.Info("Sleeping for 1 minute before updating the next pod, or completion")
+		time.Sleep(time.Minute)
+
+		// After Update(), refresh the statefulset for the next iteration (or other use).
+		err := r.Get(ctx, r.Key(), sts)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) podImageIdenticalToClusterImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string, ordinal int32,
+) error {
+	podName := fmt.Sprintf("%s-%d", sts.Name, ordinal)
+
+	var pod corev1.Pod
+	if err := r.Get(ctx, types.NamespacedName{Name: podName, Namespace: sts.Namespace}, &pod); err != nil {
+		return err
+	}
+
+	container, err := findContainer(pod.Spec.Containers, redpandaContainerName)
+	if err != nil {
+		return err
+	}
+
+	if container.Image != newImage {
+		r.logger.Info("Container image not updated to cluster image", "pod", pod.Name,
```

Comment:
This log statement can be bubble up the stack to the main reconciliation loop using returned error.

Reply:
Keep in mind that not all errors returned result in an error reaching the reconciliation loop. They might be handled on the way there. Such errors would not be logged. This is an example of such case.

Reply:
Agree with you